### PR TITLE
Take the `FORCE_COLOR` and `NO_COLOR` environnement variables into account

### PR DESCRIPTION
**Pull request by stdedos**
Support for `NO_COLOR` and `FORCE_COLOR` (and `PY_COLORS`, as an alias to `FORCE_COLOR`)
environment variables has been added. When running `pylint`, the reporter that reports
to `stdout` will be modified according to the requested mode.

The order is:

1. `NO_COLOR`
2. `FORCE_COLOR`, `PY_COLORS`
3. `--output=...`.

Signed-off-by: Stavros Ntentos <133706+stdedos@users.noreply.github.com>

-----

- [x] Document your change, if it is a non-trivial one.
  - A maintainer might label the issue ``skip-news`` if the change does not need to be in the changelog.
  - Otherwise, create a news fragment with ``towncrier create <IssueNumber>.<type>`` which will be
    included in the changelog. ``<type>`` can be one of the types defined in `./towncrier.toml`.
    If necessary you can write details or offer examples on how the new change is supposed to work.
  - Generating the doc is done with ``tox -e docs``
- [x] Relate your change to an issue in the tracker if such an issue exists (Refs #1234, Closes #1234)
- [x] Write comprehensive commit messages and/or a good description of what the PR does.
- [x] Keep the change small, separate the consensual changes from the opinionated one.
  Don't hesitate to open multiple PRs if the change requires it. If your review is so
  big it requires to actually plan and allocate time to review, it's more likely
  that it's going to go stale.
- [x] If you used multiple emails or multiple names when contributing, add your mails
      and preferred name in ``script/.contributors_aliases.json``

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :sparkles: New feature |

## Description

Fixes https://github.com/pylint-dev/pylint/issues/3995

### CODE REVIEW

**Comment by stdedos** [comment]
Any ideas of how to detect that a reporter is attached to `stdout` that works on Windows too? 😓

**Comment by hugovk** [comment]
÷1 to `FORCE_COLOR` and `NO_COLOR`. 

I suggest to leave `PY_COLORS` out because it's pytest specific. (`py` is there internal helper library and iirc it originated from there.)

https://docs.pytest.org/en/stable/reference/reference.html#envvar-PY_COLORS

File: pylint/lint/pylinter.py [hunk 1]
```diff
@@ -250,6 +261,68 @@ def _load_reporter_by_class(reporter_class: str) -> type[BaseReporter]:
...
+def _handle_force_color_no_color(reporter: list[reporters.BaseReporter]) -> None:
+    """
+    Check ``NO_COLOR``, ``FORCE_COLOR``, ``PY_COLOR`` and modify the reporter list
+    accordingly.
+
+    Rules are presented in this table:
+    +--------------+---------------+-----------------+------------------------------------------------------------+
+    | `NO_COLOR`   | `FORCE_COLOR` | `output-format` | Behavior                                                   |
+    +==============+===============+=================+============================================================+
+    | `bool: True` | `bool: True`  | colorized       | not colorized + warnings (override + inconsistent env var) |
```
Comment (stdedos) [inline pylint/lint/pylinter.py:273]: @Pierre-Sassoulas opinion? (**strictly for the `env` C0402 annotation**)

There are "other" `env` references, like e.g. https://github.com/pylint-dev/pylint/blob/14b242f66d3c8f0e3f27b05f948ae9005052e8c0/pylint/checkers/stdlib.py#L895-L932

File: doc/whatsnew/fragments/3995.feature [hunk 2]
```diff
@@ -0,0 +1,6 @@
+Support for ``NO_COLOR`` and ``FORCE_COLOR`` environment variables has been added.
+When running `pylint`, the reporter that reports to ``stdout`` will be modified according
+to the requested mode.
+The order is: ``NO_COLOR`` > ``FORCE_COLOR`` > ``--output=...``.
+
+Closes #3995 (https://github.com/pylint-dev/pylint/issues/3995).
```
Comment (DanielNoord) [inline doc/whatsnew/fragments/3995.feature:6]:
```suggestion
Closes #3995
```

File: pylint/lint/pylinter.py [hunk 3]
```diff
@@ -250,6 +260,71 @@ def _load_reporter_by_class(reporter_class: str) -> type[BaseReporter]:
 }
 
 
+def _handle_force_color_no_color(
```
Comment (DanielNoord) [inline pylint/lint/pylinter.py:263]: I changed this to not change the passed in `list` but instead just return a new list. This fixed your `mypy` issues as well as being a preferred pattern in general.

I also made it so do we loop over the list once instead of doing so twice

File: pylint/lint/pylinter.py (hunk 1 above)
Comment (DanielNoord) [reply pylint/lint/pylinter.py:273]: Fixed by adding `env` to our `custom_dict` of allowed words :)

File: tests/lint/test_pylinter.py [hunk 4]
```diff
@@ -68,3 +86,124 @@ def test_open_pylinter_prefer_stubs(linter: PyLinter) -> None:
...
+    "no_color",
+    [True, False],
+    ids=lambda no_color: f"{no_color=}",
+)
+@pytest.mark.parametrize(
+    "force_color",
+    [True, False],
+    ids=lambda force_color: f"{force_color=}",
+)
+def test_handle_force_color_no_color(
```
Comment (DanielNoord) [inline tests/lint/test_pylinter.py:146]: I'd like to have @Pierre-Sassoulas's opinion on this test. This isn't really the pattern we would normally use (e.g., use of `pytest_generate_tests`) but I'm not sure how he would prefer this to be tested.

**Review by DanielNoord — COMMENTED** [review]
Left some comments that were a bit hard to explain as in line comments, hope you don't mind.

This LGTM expect for the small change in the news entry and the test

File: pylint/lint/pylinter.py (hunk 3 above)
Comment (stdedos) [reply pylint/lint/pylinter.py:263]: > I changed this to not change the passed in `list` but instead just return a new list

I'm happy that `sub_reporters = _handle_force_color_no_color(sub_reporters)` "is a valid solution" in the code 😀 

> I also made it so do we loop over the list once instead of doing so twice

Indeed. But this is one-time setup - therefore "there is ample time to do things" 😅 

I feel that the original code (https://github.com/pylint-dev/pylint/pull/9128/files/4427cf3cdde5d2490a7498a4b9d80dfc96724fa1..51cf8d2fcbfa852449d3208b91eca9bbbbf61729) was a bit more straightforward logic/reading-wise

(I'm fine either way - just saying my view)

File: doc/whatsnew/fragments/3995.feature (hunk 2 above)
Comment (stdedos) [reply doc/whatsnew/fragments/3995.feature:6]: I have no opinion either way / Idk "how" this is rendered (neither where).

I simply prefer "links all the way", instead of having to do manual work "to see what #3995 is" (in a non-GH context)

File: tests/lint/test_pylinter.py (hunk 4 above)
Comment (stdedos) [reply tests/lint/test_pylinter.py:146]: I was a bit too quick to "ask for help", since Windows tests fail (https://github.com/pylint-dev/pylint/actions/runs/12734784829/job/35492756662?pr=9128)

I was aiming very specifically at https://github.com/pylint-dev/pylint/pull/9128#discussion_r1912476358 (which you kindly fix-up-ed yourself ❤️)

At least nowadays I have "somewhat" of Python setup in Windows and I could try to decipher this error.

Comment (stdedos) [reply tests/lint/test_pylinter.py:146]: > I'd like to have @Pierre-Sassoulas's opinion on this test. This isn't really the pattern we would normally use (e.g., use of `pytest_generate_tests`) but I'm not sure how he would prefer this to be tested.

But for the comment at-hand:

```py
@pytest.mark.parametrize(
    "no_color",
    [True, False],
    ids=lambda no_color: f"{no_color=}",
)
```

is trivial to test like so. For attaching reporters, not so much.

I could ... try to run this code once, and `@pytest.mark.parametrize` with the output of the code (ie `parameters`) instead

**Comment by stdedos** [comment]
OTOH, @DanielNoord pls note that the latest changes you committed also makes pypy tests fail? Originally (https://github.com/pylint-dev/pylint/actions/runs/12704163354) they didn't

(Note to me: Historical runs: https://github.com/pylint-dev/pylint/actions/workflows/tests.yaml?query=branch%3Aimpr%2F3995%2FNO_COLOR%2BFORCE_COLOR)

File: doc/whatsnew/fragments/3995.feature (hunk 2 above)
Comment (DanielNoord) [reply doc/whatsnew/fragments/3995.feature:6]: Since we tend to do this without links I would prefer that. The consistency makes it easier to parse these (if we ever want to) and attach the links for all of them all at once.

File: pylint/lint/pylinter.py (hunk 3 above)
Comment (DanielNoord) [reply pylint/lint/pylinter.py:263]: I personally prefer this pattern and considering you're fine with it, let's keep as is.

Thanks for sharing your opinion though, always appreciate being able to discuss different preferences in these reviews :)

**Comment by DanielNoord** [comment]
> OTOH, @DanielNoord pls note that the latest changes you committed also makes pypy tests fail? Originally (https://github.com/pylint-dev/pylint/actions/runs/12704163354) they didn't
> 
> (Note to me: Historical runs: https://github.com/pylint-dev/pylint/actions/workflows/tests.yaml?query=branch%3Aimpr%2F3995%2FNO_COLOR%2BFORCE_COLOR)

Yeah, I wanted to spend time to fix the tests as well but would first like the opinion of Pierre about the current "design" of them.
I tried to refactor them a little just know but it is really hard. There are so many magic values and fixtures that it is really difficult (for me) to follow what is actually being done. I'm sure this is testing all edge cases described in the methods table, but I'm worried that this will be a nightmare to maintain.

I would probably prefer to not use "fancy" parameterisation for these tests, but instead write them out individually. In the long run I think that is better for maintainability.

Before committing to this I would like the opinion of another maintainer though.

File: doc/whatsnew/fragments/3995.feature (hunk 2 above)
Comment (Pierre-Sassoulas) [reply doc/whatsnew/fragments/3995.feature:6]: There will be an issue with tbump parsing during release if we don't apply this so I'm applying.

File: tests/lint/test_pylinter.py (hunk 4 above)
Comment (Pierre-Sassoulas) [reply tests/lint/test_pylinter.py:146]: I feel like the test is as (or more) complicated than the code we want to test, which is bad. A middle ground would be to parametrize only one parameter and hardcode the other one ? Something like:

```python
@pytest.mark.parametrize(
    "force_color,expected",
    [(True, False)],
    ids=lambda force_color: f"{force_color=}",
)
def test_handle_force_color_no_color_true_text_reporter(...)
     monkeypatch.setenv(NO_COLOR, "1")
    reporter = TextReporter

@pytest.mark.parametrize(
    "force_color",
    [True, False],
    ids=lambda force_color: f"{force_color=}",
)
def test_handle_force_color_no_color_colorized_text_reporter(...)
     monkeypatch.setenv(NO_COLOR, "")
    reporter = ColorizedTextReporter

@pytest.mark.parametrize(
    "force_color,expected",
    [(True, False)],
    ids=lambda force_color: f"{force_color=}",
)
def test_handle_force_color_no_color_true_text_reporter(...)
     monkeypatch.setenv(NO_COLOR, "1")
     reporter = TextReporter

@pytest.mark.parametrize(
    "force_color",
    [True, False],
    ids=lambda force_color: f"{force_color=}",
)
def test_handle_force_color_no_color_colorized_text_reporter(...)
     monkeypatch.setenv(NO_COLOR, "")
    reporter = ColorizedTextReporter
```

I'm not sure if it's better than what Daniel suggest. Especially if we choose the parameter badly (we should parametrize the one that doesn't change anything because the other take precedance). Also it seems we don't actually test the color output but a warning instead?

**Comment by Pierre-Sassoulas** [comment]
This looks great, thank you for being persistent with this PR @stdedos !

File: tests/lint/test_pylinter.py (hunk 4 above)
Comment (DanielNoord) [reply tests/lint/test_pylinter.py:146]: Yes, I agree with Pierre that it would probably be better to bit more explicit and write out more of the code itself.

**Review by DanielNoord — APPROVED** [review]
This LGTM, although CI seems to disagree :(

**Comment by stdedos** [comment]
Hey @Pierre-Sassoulas, @DanielNoord,  

I know this has been taking longer than expected, and unfortunately, I need to extend the timeline further due to personal circumstances.  

If anyone feels inclined to pick this up, I’d be happy to be part of the PR - or, of course, if they prefer a fresh start, that’s totally fine too.  

I really appreciate the support you have both provided.
I’m not dropping this; it’ll just take a bit more time 🙃

**Comment by webknjaz** [comment]
May I ask you to also explicitly document whether any of `FORCE_COLOR` or `NO_COLOR` are going to treat non-color ANSI decorations like bold as color or not?
Rich, for example, enables all decorations on `FORCE_COLOR` but only disables color and not things like bold when `NO_COLOR` is also present.

**Comment by stdedos** [comment]
> May I ask you to also explicitly document whether either `FORCE_COLOR` or `NO_COLOR` treat non-color ANSI decorations (like bold) as color? Rich, for example, enables all decorations with `FORCE_COLOR`, but under `NO_COLOR`, it disables color while still allowing things like bold.

Sure, I can definitely do that. FWIW:

`NO_COLOR` - while a misnomer - is usually interpreted as:

> force-disable any non-text decoration, regardless of smart detection e.g. terminal is a tty, so decorations would normally be safe.

`FORCE_COLOR` does the opposite; 

> assume decorations are safe even if the environment says otherwise e.g. GitLab's terminal might not report correctly* but can still handle ASCII decorations (though not VT100 sequences)

And the code treats it like that - tries to replace a `ColorizedTextReporter` with a `TextReporter` if connected on the `stdout` and vice-versa.

Rich's behavior is the "first one of its kind" that I have heard of :sweat_smile:

**Comment by webknjaz** [comment]
Good to hear! I've been trying to communicate this to Rich for some while. It doesn't seem like they're interested in changing things up, despite that causing problems.

File: doc/user_guide/usage/output.rst [hunk 5]
```diff
@@ -21,6 +21,19 @@ a colorized report to stdout at the same time:
...
+''''''''''''''''''''''''''''
+The colorization of the output can also be controlled through environment
+variables. The precedence for determining output format is as follows:
+
+1. ``NO_COLOR``
+2. ``FORCE_COLOR``
+3. ``--output-format=...``
+
+Setting ``NO_COLOR`` (to any value) will disable colorized output, while
+``FORCE_COLOR`` (to any value) will enable it, overriding the
```
Comment (webknjaz) [inline doc/user_guide/usage/output.rst:34]: Probably worth being extra explicit here:
```suggestion
Setting ``NO_COLOR`` (to any non-empty value) will disable colorized output,
while ``FORCE_COLOR`` (to any non-empty value) will enable it, overriding the
```

*(should this also link https://force-color.org and https://no-color.org websites?)*

Comment (stdedos) [reply doc/user_guide/usage/output.rst]: > Read more details on the https://force-color.org and https://no-color.org websites.

?

Comment (webknjaz) [reply doc/user_guide/usage/output.rst]: @stdedos yeah, unless you want to make the var names clickable.

By the way, Sphinx allows documenting env vars. There's `.. envvar::` directive and `:envvar:` role. It's probably a good idea to declare them this way. It would then expose them to things like intersphinx, for example.

**Comment by DanielNoord** [comment]
@stdedos Do you need help with fixing these tests?

File: tests/lint/test_pylinter.py (hunk 4 above)
Comment (Pierre-Sassoulas) [reply tests/lint/test_pylinter.py]: This is resolved

**Review by Pierre-Sassoulas — APPROVED** [review]
Thank you for the initial work, I took over to merge in 4.1.0, great feature !
